File: app/src/main/java/com/mx/gillustrated/adapter/CardTypeListAdapter.java
package com.mx.gillustrated.adapter;

import java.util.List;

import com.mx.gillustrated.R;
import com.mx.gillustrated.vo.CardTypeInfo;
import com.mx.gillustrated.vo.GameInfo;

import android.content.Context;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;

public class CardTypeListAdapter extends BaseAdapter {

    private Context mcontext;
    private LayoutInflater layoutInflator;
    private List<CardTypeInfo> list;
    private DespairTouchListener mListener;

    public CardTypeListAdapter() {
    }

    public CardTypeListAdapter(Context context, List<CardTypeInfo> items) {
        mcontext = context;
        layoutInflator = LayoutInflater.from(mcontext);
        list = items;
    }

    public void setDespairTouchListener(DespairTouchListener listener) {
        mListener = listener;
    }

    @Override
    public int getCount() {
        return list.size();
    }

    @Override
    public Object getItem(int arg0) {
        return list.get(arg0);
    }

    @Override
    public long getItemId(int arg0) {
        return arg0;
    }

    @Override
    public View getView(int arg0, View convertView, ViewGroup arg2) {
        Component component = null;

        if (convertView == null) {
            convertView = layoutInflator.inflate(
                    R.layout.adapter_cardtype, null);
            component = new Component(convertView);
            convertView.setTag(component);
        } else
            component = (Component) convertView.getTag();

        final Component currentComponent = component;
        final int position = arg0;
        try {

            component.etName.setText(list.get(arg0).getName());
            component.btnSave.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    final String name = currentComponent.etName.getText().toString();
                    final int id = list.get(position).getId();
                    final int gid = list.get(position).getGameId();
                    CardTypeInfo despairInfo = new CardTypeInfo();
                    despairInfo.setId(id);
                    despairInfo.setGameId(gid);
                    despairInfo.setName(name);
                    mListener.onSaveBtnClickListener(despairInfo);
                }
            });

            component.btnDel.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    final int id = list.get(position).getId();
                    final int gid = list.get(position).getGameId();
                    CardTypeInfo cardTypeInfo = new CardTypeInfo();
                    cardTypeInfo.setId(id);
                    cardTypeInfo.setGameId(gid);
                    mListener.onDelBtnClickListener(cardTypeInfo);
                }
            });

            component.etName.setOnTouchListener(new View.OnTouchListener() {
                @Override
                public boolean onTouch(View v, MotionEvent event) {
                    if (event.getAction() == MotionEvent.ACTION_UP) {
                        //   index = position;
                    }
                    return false;
                }
            });

        } catch (Exception e) {
            e.printStackTrace();
        }

        return convertView;
    }

    public interface DespairTouchListener {
        void onSaveBtnClickListener(CardTypeInfo info);
        void onDelBtnClickListener(CardTypeInfo info);
    }

    static class Component {

        @BindView(R.id.etCardType)
        public EditText etName;

        @BindView(R.id.btnCardTypeModify)
        public Button btnSave;

        @BindView(R.id.btnCardTypeDel)
        public Button btnDel;

        public Component(View view){
            ButterKnife.bind(this, view);
        }

    }
}

File: app/src/main/java/com/mx/gillustrated/adapter/CardTypeListAdapter.kt
package com.mx.gillustrated.adapter

import com.mx.gillustrated.vo.CardTypeInfo
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import com.mx.gillustrated.activity.BaseActivity
import com.mx.gillustrated.databinding.AdapterCardtypeBinding

class CardTypeListAdapter constructor(mContext: BaseActivity, private val list: List<CardTypeInfo>) : BaseAdapter() {

    private val layoutInflater: LayoutInflater = LayoutInflater.from(mContext)
    private var mListener: DespairTouchListener? = null

    fun setDespairTouchListener(listener: DespairTouchListener) {
        mListener = listener
    }

    override fun getCount(): Int {
        return list.size
    }

    override fun getItem(arg0: Int): Any {
        return list[arg0]
    }

    override fun getItemId(arg0: Int): Long {
        return arg0.toLong()
    }

    override fun getView(arg0: Int, convertViews: View?, arg2: ViewGroup): View {
        var convertView = convertViews
        lateinit var component: AdapterCardtypeBinding

        if (convertView == null) {
            component =  AdapterCardtypeBinding.inflate(layoutInflater, arg2, false)
            convertView = component.root
            convertView.tag = component
        } else
            component = convertView.tag as AdapterCardtypeBinding

        try {

            component.etCardType.setText(list[arg0].name)
            component.btnCardTypeModify.setOnClickListener {
                val name = component.etCardType.text.toString()
                val id = list[arg0].id
                val gid = list[arg0].gameId
                val despairInfo = CardTypeInfo()
                despairInfo.id = id
                despairInfo.gameId = gid
                despairInfo.name = name
                mListener!!.onSaveBtnClickListener(despairInfo)
            }

            component.btnCardTypeDel.setOnClickListener {
                val id = list[arg0].id
                val gid = list[arg0].gameId
                val cardTypeInfo = CardTypeInfo()
                cardTypeInfo.id = id
                cardTypeInfo.gameId = gid
                mListener!!.onDelBtnClickListener(cardTypeInfo)
            }

            component.etCardType.setOnTouchListener { _, event ->
                if (event.action == MotionEvent.ACTION_UP) {
                    //   index = position;
                }
                false
            }

        } catch (e: Exception) {
            e.printStackTrace()
        }

        return convertView
    }

    interface DespairTouchListener {
        fun onSaveBtnClickListener(info: CardTypeInfo)
        fun onDelBtnClickListener(info: CardTypeInfo)
    }
}
